src/sampling.jl

```diff
@@ -488,6 +487,125 @@ function naive_wsample_norep!(a::AbstractArray, wv::WeightVec, x::AbstractArray)
     return x
 end
 
+# Weighted sampling without replacement
+#
+# Algorithm A from:
+#     Efraimidis PS, Spirakis PG (2006). "Weighted random sampling with a reservoir."
+#     Information Processing  Letters, 97 (5), 181-185. ISSN 0020-0190.
+#     doi:10.1016/j.ipl.2005.11.003.
+#     URL http://www.sciencedirect.com/science/article/pii/S002001900500298X
+#
+# Insteads of keys u^(1/w) where u = random(0,1) keys w/v where v = randexp(1) are used.
+function efraimidis_a_wsample_norep!(a::AbstractArray, wv::WeightVec, x::AbstractArray)
+    n = length(a)
+    length(wv) == n || throw(DimensionMismatch("Inconsistent lengths."))
+    k = length(x)
+
+    # calculate keys for all items
+    r = randexp(n)
+    for i in 1:n
+        @inbounds r[i] = wv.values[i]/r[i]
+    end
+
+    # return items with largest keys
+    index = sortperm(r; alg = PartialQuickSort(k), rev = true)
+    for i in 1:k
+        @inbounds x[i] = a[index[i]]
+    end
+    return x
+end
+
+# Weighted sampling without replacement
+#
+# Algorithm A-Res from:
+#     Efraimidis PS, Spirakis PG (2006). "Weighted random sampling with a reservoir."
+#     Information Processing  Letters, 97 (5), 181-185. ISSN 0020-0190.
+#     doi:10.1016/j.ipl.2005.11.003.
+#     URL http://www.sciencedirect.com/science/article/pii/S002001900500298X
+#
+# Insteads of keys u^(1/w) where u = random(0,1) keys w/v where v = randexp(1) are used.
+function efraimidis_ares_wsample_norep!(a::AbstractArray, wv::WeightVec, x::AbstractArray)
+    n = length(a)
+    length(wv) == n || throw(DimensionMismatch("Inconsistent lengths."))
+    (k = length(x)) > 0 || return x
+
+    # initialize priority queue
+    h = Array{Pair{Float64,Int}}(k)
+    @inbounds for i in 1:k
+        h[i] = (wv.values[i]/randexp() => i)
+    end
+    heapify!(h)
+
+    # set threshold
+    @inbounds T = h[1].first
```

Comment:
`T` is often used to denote a type in Julia. Unless that's the letter used in the paper, better find another one (possibly more explicit).